daemons/fenced/pacemaker-fenced.c

```diff
@@ -633,16 +635,15 @@ watchdog_device_update(void)
             rc = stonith_device_register(xml, TRUE);
             free_xml(xml);
             if (rc != pcmk_ok) {
+                exit_code = CRM_EX_FATAL;
                 crm_crit("Cannot register watchdog pseudo fence agent");
-                crm_exit(CRM_EX_FATAL);
+                stonith_shutdown(0);
```

Comment:
This means we need a return after the first call to watchdog_device_update() in update_cib_cache_cb()

Reply:
Hmm, yeah. May have to give `watchdog_device_update()` a return code or something, to indicate whether to return there. If `watchdog_device_update()` didn't call `stonith_shutdown()`, then we want to proceed in `update_cib_cache_cb()`.

Reply:
or I guess checking the exit code would work too

Reply:
Not quite, then it wouldn't get called unless `need_full_refresh()`. If we want to avoid two calls then we'll need to track whether we made the first call or not.

Reply:
I mean like

    if (need_full_refresh) {
        fencing_topology_init();
        cib_devices_update();
    }
    watchdog_device_update();

Reply:
Actually I don't think we want to ever do the first block with watchdog_device_update() and the full refresh block. Maybe:

    if (timeout_ms_saved != stonith_watchdog_timeout_ms) {
            need_full_refresh = TRUE;
    }
    if (need_full_refresh) {
        fencing_topology_init();
        cib_devices_update();
    } else {
            update_fencing_topology(event, msg);
            update_cib_stonith_devices(event, msg);
    }
    watchdog_device_update();

Not necessary for this PR though

Reply:
I haven't looked into what all the functions are doing yet. If we assume that the original logic was correct, and if the ordering of calls matters, then both of the options above would be somewhat different from the original. There's another case that sets `need_full_refresh = TRUE`: `(local_cib == NULL)`.

So in theory, we could wind up in the `else` block even if `need_full_refresh` is set to TRUE. At the end of the `else` block, we want to exit/return if `watchdog_device_update()` fails. If it succeeds, or if we took the `if` condition instead of the `else` condition, then we go on to the `if (need_full_refresh)` block.

The last `watchdog_device_update()` is probably redundant like you said if we already called it successfully in the `else` block though. If we want to make sure we only call it once, and we reorder things to do so, we need to check whether the ordering of calls matters.

Reply:
> I haven't looked into what all the functions are doing yet. If we assume that the original logic was correct,

I'm pretty sure it's not. The full refresh and the partial update should be alternatives, and not possible to both be hit.

> and if the ordering of calls matters, then both of the options above would be somewhat different from the original. There's another case that sets `need_full_refresh = TRUE`: `(local_cib == NULL)`.
> 
> So in theory, we could wind up in the `else` block even if `need_full_refresh` is set to TRUE.

Right, that's what my suggestion above is meant to avoid. If we're doing a full refresh, there's no point in applying a partial update first, because it will just get thrown away.

> At the end of the `else` block, we want to exit/return if `watchdog_device_update()` fails. If it succeeds, or if we took the `if` condition instead of the `else` condition, then we go on to the `if (need_full_refresh)` block.
> 
> The last `watchdog_device_update()` is probably redundant like you said if we already called it successfully in the `else` block though. If we want to make sure we only call it once, and we reorder things to do so, we need to check whether the ordering of calls matters.

I believe my most recent suggestion above is what should happen. We either do a full refresh or a partial update (never both), then we update the watchdog no matter what.

Reply:
Yeah, prior to 243139b2, it was impossible for `need_full_refresh` to be true in the `else` block. I don't know if the rest of the logic in `update_cib_cache_cb()` is right or not (don't want to go down that little rabbit hole), but your suggestion seems very reasonable in light of the past behavior.